### DISCUSSION TITLE
Wrong python source in activated environment on M1 in VS code terminal

### DISCUSSION THREAD

**Discussion by Aerylia** [conda-forge/miniforge#252] [Q&A] [answered]
- [x] I read [the conda-forge documentation](https://conda-forge.org/docs/user/introduction.html#how-can-i-install-packages-from-conda-forge) and could not find the solution for my problem there.

Issue:
I've installed Miniforge using [these instructions](https://developer.apple.com/metal/tensorflow-plugin/) in the standard Terminal app, using manual initialization `conda init --all` (automatic initialization gives this problem in Terminal app as well, see below).
However, when using the integrated Visual Studio code terminal, I can activate the conda (base) environment, but `which python` gives `/usr/bin/python` (Python 2.7) instead of the conda version.  

In Terminal app:
```
(base) $ which python
/Users/griendar/miniforge3/envs/env/bin/python
(base) $ python -V
Python 3.9.7
```
In Visual studio code terminal:
```
(base) $ which python
/usr/bin/python
(base) $ python -V
Python 2.7.16
```

`conda init` does not change anything either, since it should already be initialized.
<details>

```
$ conda init --all
no change     /Users/griendar/miniforge3/condabin/conda
no change     /Users/griendar/miniforge3/bin/conda
no change     /Users/griendar/miniforge3/bin/conda-env
no change     /Users/griendar/miniforge3/bin/activate
no change     /Users/griendar/miniforge3/bin/deactivate
no change     /Users/griendar/miniforge3/etc/profile.d/conda.sh
no change     /Users/griendar/miniforge3/etc/fish/conf.d/conda.fish
no change     /Users/griendar/miniforge3/shell/condabin/Conda.psm1
no change     /Users/griendar/miniforge3/shell/condabin/conda-hook.ps1
no change     /Users/griendar/miniforge3/lib/python3.9/site-packages/xontrib/conda.xsh
no change     /Users/griendar/miniforge3/etc/profile.d/conda.csh
no change     /Users/griendar/.bash_profile
no change     /Users/griendar/.zshrc
no change     /Users/griendar/.config/fish/config.fish
no change     /Users/griendar/.xonshrc
no change     /Users/griendar/.tcshrc
No action taken.
```
</details>

Note that this method of installation is the first one where I could succesfully create a virtual environment (base) where `python` was properly overwritten to another python version/location. So it might have been RNG as to that it worked in the first place.

Other failed installation attempts:
- Installation as described in the `README.md` in the Visual studio code terminal with automated initialization. This gives this same problem on both the VS code terminal and Terminal app. Even after running `conda init` in the Terminal app. Moreover, `conda activate env` tells you to run `conda init`, but as above, the call doesn't change anything.
- Installation as described on the Apple website (link above) that I used in the end (should be the same .sh file), with automated initialization in the Terminal app. This gives the same problem in the Terminal app as well. 
- Note that I came to Miniforge after stumbling on the same problems when installing Anaconda, both the graphical 64-bit MacOS installer and the command line version. The latter warned me that I wasn't running a 64-bit machine, after which I found out that Anaconda isn't supporting the M1 chip and seemingly ignoring the entire existence of it, based on lack of errors/warnings.

I verify that I have indeed a MacBook Air with an M1 chip.

For the requested additional information, I have copy pasted the output from both terminals. The `conda list` should be the same (added for completeness), but the `conda info` is different.

<br/>
In Terminal App:

Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /Users/griendar/miniforge3:
#
# Name                    Version                   Build  Channel
brotlipy                  0.7.0           py39h5161555_1003    conda-forge
ca-certificates           2021.10.8            h4653dfc_0    conda-forge
certifi                   2021.10.8        py39h2804cbe_1    conda-forge
cffi                      1.15.0           py39h52b1de0_0    conda-forge
charset-normalizer        2.0.9              pyhd8ed1ab_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
conda                     4.11.0           py39h2804cbe_0    conda-forge
conda-package-handling    1.7.3            py39h5161555_1    conda-forge
cryptography              36.0.0           py39hfb8cd70_0    conda-forge
idna                      3.1                pyhd3deb0d_0    conda-forge
libffi                    3.4.2                h3422bc3_5    conda-forge
libzlib                   1.2.11            hee7b306_1013    conda-forge
ncurses                   6.2                  h9aa5885_4    conda-forge
openssl                   1.1.1l               h3422bc3_0    conda-forge
pip                       21.3.1             pyhd8ed1ab_0    conda-forge
pycosat                   0.6.3           py39h5161555_1009    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1            py39h2804cbe_4    conda-forge
python                    3.9.7           h54d631c_3_cpython    conda-forge
python_abi                3.9                      2_cp39    conda-forge
readline                  8.1                  hedafd6a_0    conda-forge
requests                  2.26.0             pyhd8ed1ab_1    conda-forge
ruamel_yaml               0.15.80         py39h5161555_1006    conda-forge
setuptools                59.4.0           py39h2804cbe_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
sqlite                    3.37.0               h72a2b83_0    conda-forge
tk                        8.6.11               he1e0b03_1    conda-forge
tqdm                      4.62.3             pyhd8ed1ab_0    conda-forge
tzdata                    2021e                he74cb21_0    conda-forge
urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge
wheel                     0.37.0             pyhd8ed1ab_1    conda-forge
xz                        5.2.5                h642e427_1    conda-forge
yaml                      0.2.5                h642e427_0    conda-forge
zlib                      1.2.11            hee7b306_1013    conda-forge
```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
active environment : base
    active env location : /Users/griendar/miniforge3
            shell level : 1
       user config file : /Users/griendar/.condarc
 populated config files : /Users/griendar/miniforge3/.condarc
          conda version : 4.11.0
    conda-build version : not installed
         python version : 3.9.7.final.0
       virtual packages : __osx=11.6.2=0
                          __unix=0=0
                          __archspec=1=arm64
       base environment : /Users/griendar/miniforge3  (writable)
      conda av data dir : /Users/griendar/miniforge3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/osx-arm64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /Users/griendar/miniforge3/pkgs
                          /Users/griendar/.conda/pkgs
       envs directories : /Users/griendar/miniforge3/envs
                          /Users/griendar/.conda/envs
               platform : osx-arm64
             user-agent : conda/4.11.0 requests/2.26.0 CPython/3.9.7 Darwin/20.6.0 OSX/11.6.2
                UID:GID : 502732244:984178727
             netrc file : None
           offline mode : False

 ```
</details>

<br/>
In Visual studio code terminal:

Environment (<code>conda list</code>):
<details>

```
$ conda list
# packages in environment at /Users/griendar/miniforge3:
#
# Name                    Version                   Build  Channel
brotlipy                  0.7.0           py39h5161555_1003    conda-forge
ca-certificates           2021.10.8            h4653dfc_0    conda-forge
certifi                   2021.10.8        py39h2804cbe_1    conda-forge
cffi                      1.15.0           py39h52b1de0_0    conda-forge
charset-normalizer        2.0.9              pyhd8ed1ab_0    conda-forge
colorama                  0.4.4              pyh9f0ad1d_0    conda-forge
conda                     4.11.0           py39h2804cbe_0    conda-forge
conda-package-handling    1.7.3            py39h5161555_1    conda-forge
cryptography              36.0.0           py39hfb8cd70_0    conda-forge
idna                      3.1                pyhd3deb0d_0    conda-forge
libffi                    3.4.2                h3422bc3_5    conda-forge
libzlib                   1.2.11            hee7b306_1013    conda-forge
ncurses                   6.2                  h9aa5885_4    conda-forge
openssl                   1.1.1l               h3422bc3_0    conda-forge
pip                       21.3.1             pyhd8ed1ab_0    conda-forge
pycosat                   0.6.3           py39h5161555_1009    conda-forge
pycparser                 2.21               pyhd8ed1ab_0    conda-forge
pyopenssl                 21.0.0             pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1            py39h2804cbe_4    conda-forge
python                    3.9.7           h54d631c_3_cpython    conda-forge
python_abi                3.9                      2_cp39    conda-forge
readline                  8.1                  hedafd6a_0    conda-forge
requests                  2.26.0             pyhd8ed1ab_1    conda-forge
ruamel_yaml               0.15.80         py39h5161555_1006    conda-forge
setuptools                59.4.0           py39h2804cbe_0    conda-forge
six                       1.16.0             pyh6c4a22f_0    conda-forge
sqlite                    3.37.0               h72a2b83_0    conda-forge
tk                        8.6.11               he1e0b03_1    conda-forge
tqdm                      4.62.3             pyhd8ed1ab_0    conda-forge
tzdata                    2021e                he74cb21_0    conda-forge
urllib3                   1.26.7             pyhd8ed1ab_0    conda-forge
wheel                     0.37.0             pyhd8ed1ab_1    conda-forge
xz                        5.2.5                h642e427_1    conda-forge
yaml                      0.2.5                h642e427_0    conda-forge
zlib                      1.2.11            hee7b306_1013    conda-forge

```
</details>

<br/>
Details about  <code>conda</code> and system ( <code>conda info</code> ):
<details>

```
$ conda info
     active environment : base
    active env location : /Users/griendar/miniforge3
            shell level : 2
       user config file : /Users/griendar/.condarc
 populated config files : /Users/griendar/miniforge3/.condarc
          conda version : 4.11.0
    conda-build version : not installed
         python version : 3.9.7.final.0
       virtual packages : __osx=11.6.2=0
                          __unix=0=0
                          __archspec=1=arm64
       base environment : /Users/griendar/miniforge3  (writable)
      conda av data dir : /Users/griendar/miniforge3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/conda-forge/osx-arm64
                          https://conda.anaconda.org/conda-forge/noarch
          package cache : /Users/griendar/miniforge3/pkgs
                          /Users/griendar/.conda/pkgs
       envs directories : /Users/griendar/miniforge3/envs
                          /Users/griendar/.conda/envs
               platform : osx-arm64
             user-agent : conda/4.11.0 requests/2.26.0 CPython/3.9.7 Darwin/20.6.0 OSX/11.6.2
                UID:GID : 502732244:984178727
             netrc file : None
           offline mode : False
```
</details>

**Comment by hmaarrfk** [comment]
it's likely that vscode is looking for things like miniconda3 or anaconda3 and doesn't look for miniforge.

You might want to specify to vscode that it should use an interactive shell, thereby running your bashrc file instead of just the bash profile file.

**Comment by ngam** [comment]
This sometimes happens to me in vscode --- though usually when using the remote ssh stuff, and I think it basically misaligns the envs somehow. Oftentimes, for me, it is messing up `base` vs `env` or essentially insisting on keep `base` no matter what you tell it.

What I usually do to resolve this is to start the project over in vscode; and then be a little careful about selecting the interpreter when asked. Also note: the activation of the specific env doesn't always work immediately for some weird reason. You will notice when you ask it for a new terminal window (inside vscode) it goes through the motions of `conda activate` etc.

Not sure if this is helpful, but just giving my recollection of the countless times I faced this in the past --- but not recently

**Comment by ngam** [comment]
Can you screenshot the interpreter panel? Also note: https://code.visualstudio.com/docs/python/environments#_where-the-extension-looks-for-environments and https://code.visualstudio.com/docs/python/environments#_manually-specify-an-interpreter


![](https://code.visualstudio.com/assets/docs/python/environments/enter-interpreter-path.png)

**Comment by Aerylia** [comment]
So to problem is in **Command line python** in the build-in **terminal** of VS code. I had not yet gotten around to setting up any python features of VS code itself. I was just testing in the build-in bash shell.

In response to the comments, I have added the miniforge conda and correct python interpreter of VS code. Not that that should matter, since I am running on a bash shell. See pictures.

![image](https://user-images.githubusercontent.com/2073324/150522560-fa42a8a2-c806-4780-b0c4-806cdfbf8a86.png)

![image](https://user-images.githubusercontent.com/2073324/150522859-2ffc3faf-692f-4029-b655-5fe38e1ff906.png)

![image](https://user-images.githubusercontent.com/2073324/150523260-72513d0d-f669-485b-b553-31d0f2344e45.png)

**Comment by johnmdonich** [comment, marked as answer]
I have run into this problem as well.. It is a VS Code Terminal issue when using conda. 

This [issue](https://github.com/microsoft/vscode-python/issues/7607#issuecomment-543884896) in the vscode repo describes the problem. 

### TLDR

Set the vscode terminal setting `terminal.integrated.inheritEnv`: False

![image](https://user-images.githubusercontent.com/2266667/150685818-0bf6a156-b9ae-42ba-be23-d9453c9ad07f.png)

*Although I have had colleagues where this didn't work so it is not a sure thing.*

**Reply by Aerylia** [reply]
Thanks! This worked for me.

**Reply by MountainLogic** [reply]
Did not work with me under VSC 1.132.0.   

With terminal:inherit Env unchecked, stating a new terminal with 

PS G:\My Drive\Weather\WIP\Zephyr> conda activate iio
conda : The term 'conda' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify 
that the path is correct and try again.
At line:1 char:1
+ conda activate iio
+ ~~~~~
    + CategoryInfo          : ObjectNotFound: (conda:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException 

*******

From a cmd box, "conda activate iio" works just fine

**Comment by hmaarrfk** [comment]
it is clear to me that this is an issue of integration between VScode and a custom installer of conda. I think you would do well to find somebody at VSCode to help you too. however, I do not think it is a bug for us to fix. As such I've created a discussion from this to provide a space where you all can share tips among yourselves.

If you do find a concrete answer, a PR to this repo would be appreciated summarizing the findings in the README

**Comment by ngam** [comment]
Confusing 😕

I never faced the issue you're facing. I misunderstood this as using the vscode while in a project --- for that, I used to have issues with the envs switching. But VSC never lost conda altogether for me like it is clearly losing it for you
